### PR TITLE
Update for foxy

### PR DESCRIPTION
**Pull request by ruffsl**
Update build for ROS2 foxy and current Ubuntu LTS 20.04.

### CODE REVIEW

**Comment by ruffsl** [comment]
@BillWSY , do know if there are any C++ compiler options that may effected how `std::size_t` `std::uint32_t` are cast? I'm not sure how `setExcludeLabels` is used is now suddenly an issue.

https://github.com/PointCloudLibrary/pcl/blob/af3ce2530b7ae8ed083a3515168626c587a5bbcd/segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h#L253-L256

```
#26 53.06 /opt/overlay_ws/src/CogRob/omnimapper/src/organized_segmentation/organized_segmentation_tbb.cpp:519:24:   required from here
#26 53.06 /usr/include/pcl-1.10/pcl/segmentation/euclidean_cluster_comparator.h:256:13: error: no matching function for call to ‘std::set<unsigned int>::insert(std::size_t&) const’
#26 53.06   256 |             exclude_labels_->insert (i);
#26 53.06       |             ^~~~~~~~~~~~~~~
```


https://github.com/CogRob/omnimapper/blob/12138b9f50e61d31226a2eb274b17e94d94c8a08/src/organized_segmentation/organized_segmentation_tbb.cpp#L339

<details>

```
------                                                                                                                                                                                                             
 > [builder 12/13] RUN . /opt/underlay_ws/install/setup.sh &&     colcon build       --symlink-install       --mixin release ccache       --cmake-args         --no-warn-unused-cli         -DCMAKE_CXX_FLAGS="          -w":                                                                                                                                                                                                         
#26 1.105 Starting >>> omnimapper                                                                                                                                                                                  
#26 1.117 Starting >>> omnimapper_ros_msgs                                                                                                                                                                         
#26 14.08 Finished <<< omnimapper_ros_msgs [13.0s]                                                                                                                                                                 
#26 53.06 --- stderr: omnimapper
#26 53.06 ** WARNING ** io features related to openni will be disabled
#26 53.06 ** WARNING ** io features related to openni2 will be disabled
#26 53.06 ** WARNING ** io features related to pcap will be disabled
#26 53.06 ** WARNING ** io features related to png will be disabled
#26 53.06 ** WARNING ** io features related to libusb-1.0 will be disabled
#26 53.06 ** WARNING ** visualization features related to openni will be disabled
#26 53.06 ** WARNING ** visualization features related to openni2 will be disabled
#26 53.06 ** WARNING ** apps features related to openni will be disabled
#26 53.06 In file included from /opt/overlay_ws/src/CogRob/omnimapper/src/organized_segmentation/organized_segmentation_tbb.cpp:46:
#26 53.06 /usr/include/tbb/task_scheduler_init.h:21:154: note: #pragma message: TBB Warning: tbb/task_scheduler_init.h is deprecated. For details, please see Deprecated Features appendix in the TBB reference manual.
#26 53.06    21 | #pragma message("TBB Warning: tbb/task_scheduler_init.h is deprecated. For details, please see Deprecated Features appendix in the TBB reference manual.")
#26 53.06       |                                                                                                                                                          ^
#26 53.06 In file included from /opt/overlay_ws/src/CogRob/omnimapper/src/organized_segmentation/organized_segmentation_tbb.cpp:47:
#26 53.06 /usr/include/tbb/tbb_thread.h:21:145: note: #pragma message: TBB Warning: tbb/tbb_thread.h is deprecated. For details, please see Deprecated Features appendix in the TBB reference manual.
#26 53.06    21 | #pragma message("TBB Warning: tbb/tbb_thread.h is deprecated. For details, please see Deprecated Features appendix in the TBB reference manual.")
#26 53.06       |                                                                                                                                                 ^
#26 53.06 In file included from /opt/overlay_ws/src/CogRob/omnimapper/include/omnimapper/organized_segmentation/organized_segmentation_tbb.h:48,
#26 53.06                  from /opt/overlay_ws/src/CogRob/omnimapper/src/organized_segmentation/organized_segmentation_tbb.cpp:39:
#26 53.06 /usr/include/pcl-1.10/pcl/segmentation/euclidean_cluster_comparator.h: In instantiation of ‘void pcl::EuclideanClusterComparator<PointT, PointNT, PointLT>::setExcludeLabels(const std::vector<bool>&) [with PointT = pcl::PointXYZRGBA; PointNT = pcl::Normal; PointLT = pcl::Label]’:
#26 53.06 /opt/overlay_ws/src/CogRob/omnimapper/src/organized_segmentation/organized_segmentation_tbb.cpp:339:3:   required from ‘void cogrob::OrganizedSegmentationTBB<PointT>::computeClusters() [with PointT = pcl::PointXYZRGBA]’
#26 53.06 /opt/overlay_ws/src/CogRob/omnimapper/src/organized_segmentation/organized_segmentation_tbb.cpp:519:24:   required from here
#26 53.06 /usr/include/pcl-1.10/pcl/segmentation/euclidean_cluster_comparator.h:256:13: error: no matching function for call to ‘std::set<unsigned int>::insert(std::size_t&) const’
#26 53.06   256 |             exclude_labels_->insert (i);
#26 53.06       |             ^~~~~~~~~~~~~~~
#26 53.06 In file included from /usr/include/c++/9/set:61,
#26 53.06                  from /usr/include/boost/detail/container_fwd.hpp:95,
#26 53.06                  from /usr/include/boost/container_hash/extensions.hpp:22,
#26 53.06                  from /usr/include/boost/container_hash/hash.hpp:760,
#26 53.06                  from /usr/include/boost/functional/hash.hpp:6,
#26 53.06                  from /usr/include/boost/thread/detail/thread.hpp:41,
#26 53.06                  from /usr/include/boost/thread/thread_only.hpp:22,
#26 53.06                  from /usr/include/boost/thread/thread.hpp:12,
#26 53.06                  from /opt/overlay_ws/src/CogRob/omnimapper/include/omnimapper/organized_segmentation/organized_segmentation_tbb.h:45,
#26 53.06                  from /opt/overlay_ws/src/CogRob/omnimapper/src/organized_segmentation/organized_segmentation_tbb.cpp:39:
#26 53.06 /usr/include/c++/9/bits/stl_set.h:509:7: note: candidate: ‘std::pair<typename std::_Rb_tree<_Key, _Key, std::_Identity<_Tp>, _Compare, typename __gnu_cxx::__alloc_traits<_Alloc>::rebind<_Key>::other>::const_iterator, bool> std::set<_Key, _Compare, _Alloc>::insert(const value_type&) [with _Key = unsigned int; _Compare = std::less<unsigned int>; _Alloc = std::allocator<unsigned int>; typename std::_Rb_tree<_Key, _Key, std::_Identity<_Tp>, _Compare, typename __gnu_cxx::__alloc_traits<_Alloc>::rebind<_Key>::other>::const_iterator = std::_Rb_tree_const_iterator<unsigned int>; std::set<_Key, _Compare, _Alloc>::value_type = unsigned int]’ <near match>
#26 53.06   509 |       insert(const value_type& __x)
#26 53.06       |       ^~~~~~
#26 53.06 /usr/include/c++/9/bits/stl_set.h:509:7: note:   passing ‘const std::set<unsigned int>*’ as ‘this’ argument discards qualifiers
#26 53.06 /usr/include/c++/9/bits/stl_set.h:518:7: note: candidate: ‘std::pair<typename std::_Rb_tree<_Key, _Key, std::_Identity<_Tp>, _Compare, typename __gnu_cxx::__alloc_traits<_Alloc>::rebind<_Key>::other>::const_iterator, bool> std::set<_Key, _Compare, _Alloc>::insert(std::set<_Key, _Compare, _Alloc>::value_type&&) [with _Key = unsigned int; _Compare = std::less<unsigned int>; _Alloc = std::allocator<unsigned int>; typename std::_Rb_tree<_Key, _Key, std::_Identity<_Tp>, _Compare, typename __gnu_cxx::__alloc_traits<_Alloc>::rebind<_Key>::other>::const_iterator = std::_Rb_tree_const_iterator<unsigned int>; std::set<_Key, _Compare, _Alloc>::value_type = unsigned int]’ <near match>
#26 53.06   518 |       insert(value_type&& __x)
#26 53.06       |       ^~~~~~
#26 53.06 /usr/include/c++/9/bits/stl_set.h:518:7: note:   passing ‘const std::set<unsigned int>*’ as ‘this’ argument discards qualifiers
#26 53.06 /usr/include/c++/9/bits/stl_set.h:546:7: note: candidate: ‘std::set<_Key, _Compare, _Alloc>::iterator std::set<_Key, _Compare, _Alloc>::insert(std::set<_Key, _Compare, _Alloc>::const_iterator, const value_type&) [with _Key = unsigned int; _Compare = std::less<unsigned int>; _Alloc = std::allocator<unsigned int>; std::set<_Key, _Compare, _Alloc>::iterator = std::_Rb_tree_const_iterator<unsigned int>; std::set<_Key, _Compare, _Alloc>::const_iterator = std::_Rb_tree_const_iterator<unsigned int>; std::set<_Key, _Compare, _Alloc>::value_type = unsigned int]’
#26 53.06   546 |       insert(const_iterator __position, const value_type& __x)
#26 53.06       |       ^~~~~~
#26 53.06 /usr/include/c++/9/bits/stl_set.h:546:7: note:   candidate expects 2 arguments, 1 provided
#26 53.06 /usr/include/c++/9/bits/stl_set.h:551:7: note: candidate: ‘std::set<_Key, _Compare, _Alloc>::iterator std::set<_Key, _Compare, _Alloc>::insert(std::set<_Key, _Compare, _Alloc>::const_iterator, std::set<_Key, _Compare, _Alloc>::value_type&&) [with _Key = unsigned int; _Compare = std::less<unsigned int>; _Alloc = std::allocator<unsigned int>; std::set<_Key, _Compare, _Alloc>::iterator = std::_Rb_tree_const_iterator<unsigned int>; std::set<_Key, _Compare, _Alloc>::const_iterator = std::_Rb_tree_const_iterator<unsigned int>; std::set<_Key, _Compare, _Alloc>::value_type = unsigned int]’
#26 53.06   551 |       insert(const_iterator __position, value_type&& __x)
#26 53.06       |       ^~~~~~
#26 53.06 /usr/include/c++/9/bits/stl_set.h:551:7: note:   candidate expects 2 arguments, 1 provided
#26 53.06 /usr/include/c++/9/bits/stl_set.h:566:2: note: candidate: ‘template<class _InputIterator> void std::set<_Key, _Compare, _Alloc>::insert(_InputIterator, _InputIterator) [with _InputIterator = _InputIterator; _Key = unsigned int; _Compare = std::less<unsigned int>; _Alloc = std::allocator<unsigned int>]’
#26 53.06   566 |  insert(_InputIterator __first, _InputIterator __last)
#26 53.06       |  ^~~~~~
#26 53.06 /usr/include/c++/9/bits/stl_set.h:566:2: note:   template argument deduction/substitution failed:
#26 53.06 In file included from /opt/overlay_ws/src/CogRob/omnimapper/include/omnimapper/organized_segmentation/organized_segmentation_tbb.h:48,
#26 53.06                  from /opt/overlay_ws/src/CogRob/omnimapper/src/organized_segmentation/organized_segmentation_tbb.cpp:39:
#26 53.06 /usr/include/pcl-1.10/pcl/segmentation/euclidean_cluster_comparator.h:256:13: note:   candidate expects 2 arguments, 1 provided
#26 53.06   256 |             exclude_labels_->insert (i);
#26 53.06       |             ^~~~~~~~~~~~~~~
#26 53.06 In file included from /usr/include/c++/9/set:61,
#26 53.06                  from /usr/include/boost/detail/container_fwd.hpp:95,
#26 53.06                  from /usr/include/boost/container_hash/extensions.hpp:22,
#26 53.06                  from /usr/include/boost/container_hash/hash.hpp:760,
#26 53.06                  from /usr/include/boost/functional/hash.hpp:6,
#26 53.06                  from /usr/include/boost/thread/detail/thread.hpp:41,
#26 53.06                  from /usr/include/boost/thread/thread_only.hpp:22,
#26 53.06                  from /usr/include/boost/thread/thread.hpp:12,
#26 53.06                  from /opt/overlay_ws/src/CogRob/omnimapper/include/omnimapper/organized_segmentation/organized_segmentation_tbb.h:45,
#26 53.06                  from /opt/overlay_ws/src/CogRob/omnimapper/src/organized_segmentation/organized_segmentation_tbb.cpp:39:
#26 53.06 /usr/include/c++/9/bits/stl_set.h:578:7: note: candidate: ‘void std::set<_Key, _Compare, _Alloc>::insert(std::initializer_list<_Tp>) [with _Key = unsigned int; _Compare = std::less<unsigned int>; _Alloc = std::allocator<unsigned int>]’
#26 53.06   578 |       insert(initializer_list<value_type> __l)
#26 53.06       |       ^~~~~~
#26 53.06 /usr/include/c++/9/bits/stl_set.h:578:43: note:   no known conversion for argument 1 from ‘std::size_t’ {aka ‘long unsigned int’} to ‘std::initializer_list<unsigned int>’
#26 53.06   578 |       insert(initializer_list<value_type> __l)
#26 53.06       |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
#26 53.06 make[2]: *** [CMakeFiles/omnimapper.dir/build.make:219: CMakeFiles/omnimapper.dir/src/organized_segmentation/organized_segmentation_tbb.cpp.o] Error 1
#26 53.06 make[2]: *** Waiting for unfinished jobs....
#26 53.06 make[1]: *** [CMakeFiles/Makefile2:107: CMakeFiles/omnimapper.dir/all] Error 2
#26 53.06 make: *** [Makefile:141: all] Error 2
#26 53.06 ---
#26 53.06 Failed   <<< omnimapper	[ Exited with code 2 ]
#26 53.11 [Processing: omnimapper]
#26 53.11 
#26 53.11 Summary: 1 package finished [52.1s]
#26 53.11   1 package failed: omnimapper
#26 53.11   1 package had stderr output: omnimapper
#26 53.11   1 package not processed
------
failed to solve with frontend dockerfile.v0: failed to build LLB: executor failed running [/bin/sh -c . $UNDERLAY_WS/install/setup.sh &&     colcon build       --symlink-install       --mixin $OVERLAY_MIXINS       --cmake-args         --no-warn-unused-cli         -DCMAKE_CXX_FLAGS="          -w"]: runc did not terminate sucessfully
```

</details>

In 46f3b90 , I just updated the organized segmentation use of plane_labels to match that in the latest version of the original example @atrevor added to upstream to the PCL library apps.

* https://github.com/PointCloudLibrary/pcl/blob/70047873577e4d9b6d6d6e675eafaff3216492bc/apps/src/organized_segmentation_demo.cpp
* https://github.com/PointCloudLibrary/pcl/blob/70047873577e4d9b6d6d6e675eafaff3216492bc/apps/include/pcl/apps/organized_segmentation_demo.h

**Comment by BillWSY** [comment]
I doubt there would be such an option. There are two options.

(1) Patch PCL code. It is straightforward and safe to static_cast between size_t and uint32_t.
(2) As the code indicates
```
deprecated("use setExcludeLabels(const ExcludeLabelSetConstPtr &) instead")
```
So we can construct a `ExcludeLabelSet` and pass it to `setExcludeLabels(const ExcludeLabelSetConstPtr &)`

File: src/organized_segmentation/organized_segmentation_tbb.cpp [hunk 1]
```diff
@@ -322,14 +322,15 @@ void OrganizedSegmentationTBB<PointT>::computeClusters() {
 
   clust_output_labels_ = LabelCloudPtr(new LabelCloud());
 
-  std::vector<bool> plane_labels;
-  plane_labels.resize(clust_input_label_indices_->size(), false);
+  typename pcl::EuclideanClusterComparator<PointT, pcl::Label>::ExcludeLabelSetPtr
```
Comment (BillWSY) [inline src/organized_segmentation/organized_segmentation_tbb.cpp:325]: consider `boost::make_shared`

```
auto plane_labels = boost::make_shared<pcl::EuclideanClusterComparator<PointT, pcl::Label>::ExcludeLabelSet>();
```

**Review by BillWSY — APPROVED** [review]

File: src/organized_segmentation/organized_segmentation_tbb.cpp (hunk 1 above)
Comment (ruffsl) [reply src/organized_segmentation/organized_segmentation_tbb.cpp:325]: I think this causes some issues given the type seems to be a dependent name.

```
omnimapper/src/organized_segmentation/organized_segmentation_tbb.cpp:325:30: error: dependent-name ‘pcl::EuclideanClusterComparator<PointT, pcl::Label>::ExcludeLabelSet’ is parsed as a non-type, but instantiation yields a type
  325 |   auto plane_labels = boost::make_shared<pcl::EuclideanClusterComparator<PointT, pcl::Label>::ExcludeLabelSet>();
      |                       ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Pull request merged**